waspc/src/Wasp/Generator/NpmInstall.hs

```diff
@@ -100,34 +109,81 @@ loadInstalledFullStackNpmDependencies dstDir = do
+      threadDelay $ secToMicroSec 5
+      reportPeriodically (if hasLessThan2Elems messages then messages else drop 1 messages)
+    secToMicroSec = (* 1000000)
+    hasLessThan2Elems = null . drop 1
+    allPossibleMessages =
+      [ "Still installing npm dependencies!",
+        "Installation going great - we'll get there soon!",
+        "The installation is taking a while, but we'll get there!",
+        "Yup, still not done installing.",
+        "We're getting closer and closer, everything will be installed soon!",
+        "Still waiting for the installation to finish? You should! We got too far to give up now!",
+        "You've been waiting so patiently, just wait a little longer (for the installation to finish)..."
+      ]
+
+installNpmDependenciesAndReport :: Job -> Chan JobMessage -> JobType -> IO ExitCode
+installNpmDependenciesAndReport installF chan jobType = do
+  writeChan chan $ J.JobMessage {J._data = J.JobOutput "Starting npm install\n" J.Stdout, J._jobType = jobType}
+  result <- installF chan `race` reportInstallationProgress chan jobType
+  case result of
+    Left exitCode -> return exitCode
+    Right _ -> error "This should be impossible"
+
 -- Run the individual `npm install` commands for both server and webapp projects
 -- It runs these concurrently, collects the output produced by these commands
 -- to pass them along to IO with a prefix
-installNpmDependencies :: Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
-installNpmDependencies projectDir = do
-  chan <- newChan
-  let runSetupJobs =
-        ServerSetup.installNpmDependencies projectDir chan
-          `concurrently` WebAppSetup.installNpmDependencies projectDir chan
-  (_, result) <- concurrently (handleJobMessages chan) runSetupJobs
-  case result of
-    (ExitSuccess, ExitSuccess) -> return $ Right ()
-    exitCodes -> return $ Left $ setupFailedMessage exitCodes
+installNpmDependencies :: Path' Abs (Dir WaspProjectDir) -> Path' Abs (Dir ProjectRootDir) -> IO (Either String ())
+installNpmDependencies projectDir dstDir = do
+  messagesChan <- newChan
+  (_, exitCode) <-
+    concurrently
+      (handleProjectInstallMessage messagesChan)
+      (installNpmDependenciesAndReport (SdkGenerator.installNpmDependencies projectDir) messagesChan J.Wasp)
+  case exitCode of
+    ExitFailure code -> return $ Left $ "Project setup failed with exit code " ++ show code ++ "."
+    _ -> do
+      let handleMessagesJob = handleJobMessages messagesChan
+      let runSetupJobs =
+            concurrently
+              (installNpmDependenciesAndReport (ServerSetup.installNpmDependencies dstDir) messagesChan J.Server)
+              (installNpmDependenciesAndReport (WebAppSetup.installNpmDependencies dstDir) messagesChan J.WebApp)
+      (_, results) <- concurrently handleMessagesJob runSetupJobs
+      case results of
+        (ExitSuccess, ExitSuccess) -> return $ Right ()
+        exitCodes -> return $ Left $ setupFailedMessage exitCodes
```

Comment:
Possibly nicer:

```suggestion
      (handleMessagesJob `concurrently` runSetupJobs) >>= \case
        (ExitSuccess, ExitSuccess) -> return $ Right ()
        exitCodes -> return $ Left $ setupFailedMessage exitCodes
```

You could also maybe extract that return in front of case: `return . \case`, I think that should compile.